langchain/prompt.py

```diff
@@ -126,3 +136,110 @@ def from_examples(
         example_str = example_separator.join(examples)
         template = prefix + example_str + suffix
         return cls(input_variables=input_variables, template=template)
+
+
+class DynamicPrompt(BaseModel, BasePrompt):
+    r"""Schema to represent a dynamic prompt for an LLM.
+
+    Example:
+        .. code-block:: python
+
+            from langchain import DynamicPrompt
+            dynamic_prompt = DynamicPrompt(
+                examples=["Say hi. Hi", "Say ho. Ho"],
+                example_separator="\n\n",
+                prefix="",
+                suffix="\n\nSay {foo}"
+                input_variables=["foo"],
+                max_length=200,
+                get_text_length=word_count
+            )
+    """
+
+    examples: List[str]
+    """A list of the examples that the prompt template expects."""
+
+    example_separator: str
+    """Example separator, e.g. \n\n, for the dynamic prompt creation."""
+
+    input_variables: List[str]
+    """A list of the names of the variables the prompt template expects."""
+
+    prefix: str
+    """Prefix for the prompt."""
+
+    suffix: str
+    """Suffix for the prompt."""
+
+    template_format: str = "f-string"
+    """The format of the prompt template. Options are: 'f-string'."""
+
+    get_text_length: Callable = count_words
+    """Function to measure prompt length. Defaults to word count."""
+
+    max_length: int = 2048
+    """Max length for the prompt, beyond which examples are cut."""
+
+    class Config:
+        """Configuration for this pydantic object."""
+
+        extra = Extra.forbid
+
+    def format(self, **kwargs: Any) -> str:
+        """Dynamically format the prompt with the inputs.
+
+        Args:
+            kwargs: Any arguments to be passed to the prompt template.
+
+        Returns:
+            A formatted string.
+
+        Example:
+
+        .. code-block:: python
+
+            prompt.format(variable1="foo")
+        """
+
+        def return_template(example_list: List[str]) -> str:
```

Comment:
im not a huge fan of nested functions, can we make this a separate class method or is that hard for some reason?